src/rpc/net.cpp

```diff
@@ -63,9 +63,14 @@ UniValue setmaxconnections(const JSONRPCRequest& request)
             + HelpExampleRpc("setmaxconnections", minConnCount)
         );
     else
-        newMaxCount = request.params[0].get_int();
+    {
+        // this is really gross, but it avoids signedness conversions later
+        int incomingCount = request.params[0].get_int();
+        if (incomingCount > 0)
+            newMaxCount = incomingCount;
+    }
 
-    if (newMaxCount < MAX_ADDNODE_CONNECTIONS)
+    if (newMaxCount == 0 || newMaxCount < MAX_ADDNODE_CONNECTIONS)
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Error: maxconnectioncount must be >= " + minConnCount);
```

Comment:
I was thinking this may be cleaner as:

```suggestion
    int incomingCount = request.params[0].get_int();
    if (incomingCount < 1) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, "Error: maxconnectioncount must be >= " + minConnCount);
    }
```

and then just pass a `uint32_t(incomingCount)` at line 79?

Reply:
That's better. I'm a little reluctant to remove the second half of the conditional on line 75 because I added a test for it, but I don't feel too strongly.

I've just pushed a couple of additional commits to address this.

Reply:
I understand what you mean - I missed the fact that that was comparing to that const which is currently 8, and yeah that should probably stay.

Another issue is that the minimum isn't really 8 of course, because we later silently change it if it's under 28. I don't really mind either way, just that saying that it should be > 8 and then silently not allowing < 28 may be confusing. 🤔 I'm unsure now.

Reply:
I thought about that too. In the `init.cpp` code, we're only evaluating the number of max allowed incoming connections. In the net code, we're comparing against the minimum number of _all_ connections, of which max allowed incoming is a part.

We could separate out counts of all types of connections, but that's a lot of churn in the net code and a lot of runtime calculations for what seems like little benefit.

I wonder if there's a way to phrase the debug messages in the other code that makes this clearer. We'll have the same questions when I get back to the Qt code too.

Reply:
Okay so I thought about this, and we had 2 considerations in #2698:

1. We don't want to leak file descriptor information over RPC so we do not inform outside of the log when a value greater than the system maximum is requested 
2. We don't want to accept a `maxconnections` value that is lower than what we can actually get to by _evicting nodes_ once we're up and running

and then on the discussion above, we said
3. We can accept a value lower than what we can actually evict towards when _starting_ a node (but not negative values.)

So I think this would be most consistently implemented by doing:
- in `init.cpp`, we just accept anything and make sure that it is >= 0 as we do now, but with a log message as you proposed. 
- in `net.cpp`, we keep the checks as they are but we improve the log messages where needed
- in `rpc/net.cpp`, we check the minimum against the awesome new function `CConnman::GetMinConnections()` that we introduced specifically for making that check - from when we were looking at Qt

Reply:
I'm on board with all of this, but I wonder if the change for `rpc/net.cpp` shows that `CConnman::GetMinConnections()` has the wrong semantics. If I change to compare against that, the lower bound for new max connections becomes 28 instead of 8 (poke at the `setmaxconnections.py` test for proof).

Ultimately I think we're trying to balance two things:

 * a single node needs to maintain _n_ connections to work at all
 * a single node can have a maximum threshold of incoming connections

Maybe `CConnman::GetMinConnections()` shouldn't be trying to give us numbers for both of those. I'll keep thinking about this though.

Reply:
What if the problem is that we're treating max connections one way from the user's point of view (set the number of maximum incoming connections) but another way within the network loop (what's the total number of allowed connections, regardless of inbound, outbound, evictable, etc)?

If this is true, then we should calculate the latter separately in `src/net.cpp`, using the former as a portion of that count. That probably means separate methods to calculate these values (so we calculate them in a single place), and a validation method that takes a new proposed max incoming collection count and compares that against the threshold of total allowed connections to see if there are too many proposed.

I think that gets us closer to what this code would look like if we were creating it afresh. What do you think?